dali/pipeline/operator/sequence_operator.h

```diff
@@ -0,0 +1,451 @@
+      if (ws.template OutputIsType<GPUBackend>(output_idx)) {
+        SetOutputLayoutHelper<GPUBackend>(ws, output_idx, layout_prefix);
+      } else {
+        SetOutputLayoutHelper<CPUBackend>(ws, output_idx, layout_prefix);
+      }
+    }
+  }
+
+  virtual void CoalesceOutputShapes(std::vector<OutputDesc> &output_desc,
+                                    const workspace_t<Backend> &ws) {
+    for (size_t output_idx = 0; output_idx < output_desc.size(); output_idx++) {
+      const auto &expand_desc = GetOutputExpandDesc(ws, output_idx);
+      const auto &shape = output_desc[output_idx].shape;
+      DALI_ENFORCE(shape.num_samples() == expand_desc.NumElements(),
+                   make_string("Unexpected number of frames inferred in the operator for output ",
+                               output_idx, ". Expected ", expand_desc.NumElements(),
+                               " but the operator returned ", shape.num_samples(), "."));
+      output_desc[output_idx].shape = fold_outermost_like(shape, expand_desc.ExpandedShape());
+    }
+  }
+
+  virtual bool ProcessOutputDesc(std::vector<OutputDesc> &output_desc,
+                                 const workspace_t<Backend> &ws, bool is_inferred) {
+    if (is_inferred && expand_) {
+      CoalesceOutputShapes(output_desc, ws);
+    }
+    return is_inferred;
+  }
+
+  virtual void ExpandArgument(const workspace_t<Backend> &ws, const std::string &arg_name,
+                             const TensorVector<CPUBackend> &arg_tensor) {
+    const auto &expand_desc = GetArgExpandDesc(arg_name);
+    auto expanded_arg = ExpandArgumentLike(arg_tensor, arg_name, expand_desc);
+    auto expanded_handle = std::make_shared<TensorVector<CPUBackend>>(std::move(expanded_arg));
+    AddArgumentInputHelper(arg_name, expanded_handle);
+  }
+
+  virtual void ExpandArguments(const workspace_t<Backend> &ws) {
+    for (const auto &arg_input : ws) {
+      auto shared_tvec = arg_input.second.tvec;
+      assert(shared_tvec);
+      ExpandArgument(ws, arg_input.first, *shared_tvec);
+    }
+  }
+
+  virtual const ExpandDesc &GetOutputExpandDesc(const workspace_t<Backend> &ws, int output_idx) {
+    (void)ws;
+    return GetInputExpandDesc(output_idx);
+  }
+
+  virtual const ExpandDesc &GetArgExpandDesc(const std::string &arg_name) {
+    (void)arg_name;
+    return GetInputExpandDesc(0);
+  }
+
+  bool HasPerFrameArgInput(const workspace_t<Backend> &ws) {
+    for (const auto &arg_input : ws) {
+      auto shared_tvec = arg_input.second.tvec;
+      assert(shared_tvec);
+      if (IsPerFrame(*shared_tvec)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  bool IsPerFrame(const TensorVector<CPUBackend> &arg_tensor) {
+    const auto &layout = arg_tensor.GetLayout();
+    return layout.size() > 0 && layout[0] == 'F';
+  }
+
+  template <typename InputBackend>
+  ws_input_t<InputBackend> ExpandInputHelper(const workspace_t<Backend> &ws, int input_idx,
```

Comment:
```suggestion
  ws_input_t<InputBackend> ExpandInput(const workspace_t<Backend> &ws, int input_idx,
```
?
Also: I understand that currently expanding some inputs and broadcasting others would require copying samples (in case of GPU operators), but perhaps we should do it regardless and optimize later. Currently I don't see that there is an option to broadcast some inputs and expand others and that could be quite useful.

Reply:
Right, there is no "broadcast-helper" for inputs now. I think it could be a follow-up or even added when the use-case shows up. What do you think?

Could you elaborate on the copying argument? I am not sure if I follow.

Reply:
As mentioned elsewhere - added the logic to handle multi-inputs this way, but left the broadcasting itself with DAIL_FAIL placeholder for now and do it as follow up if that is OK.